bot/helpers/db.py

```diff
@@ -0,0 +1,195 @@
+
+        return {col: val for col, val in row.items()}
+
+    async def _update_row(self, table, obj, **data):
+        """ Generic method to update table row by object id. """
+        col_vals = ',\n    '.join(f'{col} = {val}' for col, val in data.items())
+        ret_vals = ',\n    '.join(data)
+        statement = (
+            f'UPDATE {table}\n'
+            f'SET {col_vals}\n'
+            'WHERE\n'
+            '    id = $1\n'
+            f'RETURNING {ret_vals};'
+        )
+
+        async with self.pool.acquire() as connection:
+            async with connection.transaction():
+                updated_vals = await connection.fetch(statement, obj.id)
+
+        return {col: val for rec in updated_vals for col, val in rec.items()}
+
+    async def insert_guilds(self, *guilds):
+        """ Add a list of guilds into the guilds table and return the ones successfully added. """
+        rows = [(guild.id, None, None, None) for guild in guilds]
+        statement = (
+            'INSERT INTO guilds (id)\n'
+            '    (SELECT id FROM unnest($1::guilds[]))\n'
+            '    ON CONFLICT (id) DO NOTHING\n'
+            '    RETURNING id;'
+        )
+
+        async with self.pool.acquire() as connection:
+            async with connection.transaction():
+                inserted = await connection.fetch(statement, rows)
+
+        return self._get_record_attrs(inserted, 'id')
+
+    async def delete_guilds(self, *guilds):
+        """ Remove a list of guilds from the guilds table and return the ones successfully removed. """
+        delete_ids = [guild.id for guild in guilds]
+        statement = (
+            'DELETE FROM guilds\n'
+            '    WHERE id::BIGINT = ANY($1::BIGINT[])\n'
+            '    RETURNING id;'
+        )
+
+        async with self.pool.acquire() as connection:
+            async with connection.transaction():
+                deleted = await connection.fetch(statement, delete_ids)
+
+        return self._get_record_attrs(deleted, 'id')
+
+    async def sync_guilds(self, *guilds):
+        """ Synchronizes the guilds table with the guilds in the bot. """
+        insert_rows = [(guild.id, None, None, None) for guild in guilds]
+        not_delete_ids = [guild.id for guild in guilds]
+        insert_statement = (
+            'INSERT INTO guilds (id)\n'
+            '    (SELECT id FROM unnest($1::guilds[]))\n'
+            '    ON CONFLICT (id) DO NOTHING\n'
+            '    RETURNING id;'
+        )
+        delete_statement = (
+            'DELETE FROM guilds\n'
+            '    WHERE id::BIGINT != ALL($1::BIGINT[])\n'
+            '    RETURNING id;'
+        )
+
+        async with self.pool.acquire() as connection:
+            async with connection.transaction():
+                inserted = await connection.fetch(insert_statement, insert_rows)
+                deleted = await connection.fetch(delete_statement, not_delete_ids)
```

Comment:
Does it matter what order these are called in?

Reply:
Not really. Was trying to think of the best way to update the guilds table to match a list of guilds. Not sure if this is best.